### DISCUSSION TITLE
What things do Fief Workers do?

### DISCUSSION THREAD

**Discussion by ScottFred** [fief-dev/fief#309] [Q&A] [answered]
I've seen examples of Fief infrastructure that include a Fief Worker in the Docker Compose.  
1. How do I know if I need one?
2. What are the pros/cons of having a Fief Worker?

**Comment by frankie567** [comment, marked as answer]
Fief *requires* to have a worker running to work properly. This is where happens all the background logic, like email sending, roles/permissions updates, tokens cleanup, etc.

To ease things and get started quickly, our Docker image starts a `supervisord` process which will spawn all the required processes (the server, the worker and a Redis server):

https://github.com/fief-dev/fief/blob/9296856e09e9a50993ed5c28dda8c0d0b8a97e18/docker/supervisord.conf#L1-L17

---

While convenient for development and test purposes, it's generally admitted that it's not a good practice to have a Docker container run several processes. That's why our [Docker Compose example](https://docs.fief.dev/self-hosting/deployment/docker-compose/#docker-compose) recommends to have separate containers for each parts: a server, a worker and a Redis (and an SQL database).

In terms of functionality, it doesn't change anything. It's just better for resilience and monitoring in production. Typically, you could independently scale the server and/or worker processes to handle more workload.

**Reply by ScottFred** [reply]
Great!  Thanks for the extra details!
